google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,195 @@
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_COUNT = 2000
+        MAX_SIZE = 10000000
+
+        attr_reader :service, :queue
+        attr_reader :max_size, :max_count
+        attr_reader :thread_pool, :publish_task
+
+        def initialize service, max_size: MAX_SIZE, max_count: MAX_COUNT
+          @service = service
+          @max_size = max_size
+          @max_count = max_count
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
+          )
+          @publish_task = Concurrent::TimerTask.new(execution_interval: 0.25) do
+            flush
+          end
+
+          # init MonitorMixin
+          super()
+        end
+
+        def publish topic_name, data = nil, attributes = {}, &block
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_count: max_count,
+                                             max_size: max_size)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @publish_task.execute
+        end
+
+        def stop
+          flush
+          @publish_task.shutdown
+        end
+
+        def started?
+          @publish_task.running?
+        end
+
+        def stopping?
+          @publish_task.shuttingdown?
+        end
+
+        def stopped?
+          @publish_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @publish_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
+            attributes = data
+            data = nil
+          end
+
+          # Convert IO-ish objects to strings
+          if data.respond_to?(:read) && data.respond_to?(:rewind)
+            data.rewind
+            data = data.read
+          end
+          # Convert data to encoded byte array to match the protobuf defn
+          data = String(data).force_encoding("ASCII-8BIT")
```

Comment:
FWIW, we understand the difference between `#force_encoding` and `#encode!`, and we chose `#force_encoding` to try to preserve the data that the user presents. Before we make any changes to how the method encodes the data, can we get consensus on whether we should continue to be liberal in allowing non-binary data and encoding it to binary, or should we be conservative and force users to encode the data before calling `#publish`?